src/contract/contract.rs

```diff
@@ -31,18 +31,38 @@ use std::str::FromStr;
 
 use amplify::confinement::{LargeOrdMap, LargeOrdSet, SmallVec, TinyOrdMap};
 use amplify::hex;
-use bp::seals::txout::TxoSeal;
 use bp::{Outpoint, Txid};
 use strict_encoding::{StrictDecode, StrictDumb, StrictEncode};
 
 use crate::contract::contract::WitnessOrd::OffChain;
 use crate::{
     Assign, AssignmentType, Assignments, AssignmentsRef, ContractId, ExposedSeal, ExposedState,
     Extension, Genesis, GlobalStateType, OpId, Operation, RevealedAttach, RevealedData,
-    RevealedValue, SchemaId, SealWitness, SubSchema, Transition, TypedAssigns, VoidState,
-    LIB_NAME_RGB,
+    RevealedValue, SchemaId, SealDefinition, SealWitness, SubSchema, Transition, TypedAssigns,
+    VoidState, LIB_NAME_RGB,
 };
 
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Display)]
+#[derive(StrictType, StrictDumb, StrictEncode, StrictDecode)]
+#[strict_type(lib = LIB_NAME_RGB, tags = custom)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate", rename_all = "camelCase")
+)]
+#[display(inner)]
+#[non_exhaustive]
+pub enum Output {
+    #[strict_type(tag = 0x00)]
+    Bitcoin(Outpoint),
+    #[strict_type(tag = 0x01)]
+    Liquid(Outpoint),
+    #[strict_type(tag = 0x10)]
+    Abraxas,
+    #[strict_type(tag = 0x11, dumb)]
+    Prime,
+}
```

Comment:
Nice!

One question, how does this handle testnets?

Reply:
Testnets are defined by `chain` field in genesis. The difference is that with `Layer1Set` a contract can operate multiple chain types, while testnet is orthogonal to that. I think it worth renaming `chain` to `testnet` and make it bool.

Reply:
That might make sense.

Reply:
Shouldn't we consider Testnet just like any other layer 1? Also making testnet a bool makes it hard to distinguish between multiple testnets (eg signet, regtest etc).

Reply:
The idea is the following: the asset/contract is always either mainnet - or testnet. I.e. we must prevent mistakes like somebody creates a contract working on both testnet and mainnet since the asset can move from one chain to another - and back - and the security of one of the chain is clearly poorer than the other. Thus, "testnets" are not just another layer 1 for us, but there are two distinct sets of layer 1: testnet or mainnet. If an asset works with let's say both bitcoin and liquid, it is either on bitcoin and liquid mainnnet - or on bitcoin and liquid testnet (but never on bitcoin mainnet and testnet, or liquid testnet and bitcoin mainnet).

Regarding different testnets - I do not see how to support them without breaking the invariant requirement above. Signet/regtest exists only on Bitcoin and not on Liquid. Thus, like addresses do not always distinguish testnets (signet and testnet3 has the same address) I think the distinction is not that important and users while testing should just agree on specific testnet used out-of-band.

Reply:
I think it is reasonable to expect contract issuers and wallets to understand the different security that mainnet and testnet offer, just like we expect them to understand the massive trade-off between Bitcoin and Liquid. Supporting multiple Bitcoin testnets to me seems a more useful feature, especially for people that want to test cross chain assets. Moreover Bitmask already supports both testnet and signet, so to make it easy to distinguish between testnets is something that would benefit users already today.

Reply:
Yes, the support for multiple testnets are there anyway. It is just provided by the resolver and invoces - and not consensus. What consensus differenciate is distinct transaction formats (like between Liquid and Bitcoin).

Reply:
I propose to move the discussion into https://github.com/RGB-WG/rgb-core/pull/189 which is still umerged